src/ucp/core/ucp_am.c

```diff
@@ -280,6 +280,7 @@ ucp_am_zcopy_pack_user_header(ucp_request_t *req)
 
     ucp_am_pack_user_header(reg_desc + 1, req);
     req->send.msg_proto.am.reg_desc = reg_desc;
+    req->flags                     |= UCP_REQUEST_FLAG_AM_HDR_PACKED;
```

Comment:
this phase should be done when AM zcopy send is started, not during progress. then, we won't need to save *header in the request at all

Reply:
ok, but header is needed for bcopy and short

Reply:
right, still need union, but each proto would use other field of it (can comment this in the struct)